File: wapp/js/views/lmanager/Misc.js
/*
Copyright 2012, Nabil SEFRIOUI

This file is part of Skyproc.

Skyproc is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of
the License, or any later version.

Skyproc is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with Skyproc. If not, see <http://www.gnu.org/licenses/>.
*/

// NS
Ext.ns('Sp.lmanager');

Sp.lmanager.isInTheAir = function(loadRec) {
    return ['P', 'B'].indexOf(loadRec.data.state) == -1;
}

Sp.lmanager.slotsSorter_get_name = function(r) {
    if (r.data.person) {
        return Sp.ui.misc.formatFullname({
            data : r.data.person
        }, Data.me.data.name_order, true).toLowerCase();
    } else if (r.data.phantom) {
        return r.data.phantom.name.toLowerCase();
    } else if (r.data.worker) {
        var name;
        Data.locations.each(function(l) {
            var w = l.Workers().getById(r.data.worker);
            if (w) {
                name = w.data.name.toLowerCase();
                return false;
            }
        });
        return name;
    }

};

Sp.lmanager.slotsSorter_compare_names = function(n1, n2) {
    var name_rank = 0;
    if (n1 || n2) {
        if (!n1) {
            name_rank = 1;
        } else if (!n2) {
            name_rank = -1;
        } else {
            if (n1 > n2) {
                name_rank = 1;
            } else if (n1 < n2) {
                name_rank = -1;
            }
        }
    }
    return name_rank;
};

Sp.lmanager.slotsSorter = function(r1, r2) {

    var get_name = Sp.lmanager.slotsSorter_get_name;
    var compare_names = Sp.lmanager.slotsSorter_compare_names;

    // get order mode
    if (r1.store && r1.store.slotsOrder) {
        var order = r1.store.slotsOrder;
    } else {
        var order = 'exit';
    }
    var rank = 0;

    var r1_empty = (!r1.data.item && !r1.data.person && !r1.data.phantom && !r1.data.worker);
    var r2_empty = (!r2.data.item && !r2.data.person && !r2.data.phantom && !r2.data.worker);

    if (r1_empty && r2_empty) {
        order = 'exit';
    } else if (r1_empty || r2_empty) {// empty slots are always shown last
        if (r1_empty) {
            return 1;
        } else {
            return -1;
        }
    }

    // sort by exits
    if (order == 'exit') {
        var r1_exit = r1.data.exit_order;
        var r2_exit = r2.data.exit_order;
        if (Ext.isNumber(r1_exit) || Ext.isNumber(r2_exit)) {
            if (!Ext.isNumber(r1_exit)) {
                rank = 1;
            } else if (!Ext.isNumber(r2_exit)) {
                rank = -1;
            } else {
                if (r1_exit > r2_exit) {
                    rank = 1;
                } else if (r1_exit < r2_exit) {
                    rank = -1;
                } else {// same exit order
                    // sort workers by role order
                    if (r1.data.worker_type && r2.data.worker_type) {
                        var wt1 = Data.workerTypes.getById(r1.data.worker_type);
                        var wt2 = Data.workerTypes.getById(r2.data.worker_type);
                        if (wt1.data.order_index > wt2.data.order_index) {
                            rank = 1;
                        } else if (wt1.data.order_index < wt2.data.order_index) {
                            rank = -1;
                        }
                        // sort no workers by name
                    } else if (!r1.data.worker_type && !r2.data.worker_type) {
                        rank = compare_names(get_name(r1), get_name(r2));
                        // workers after person/phantom
                    } else {
                        rank = r1.data.worker_type ? 1 : -1;
                    }
                }
            }
        }
    } else if (order == 'name') {
        rank = compare_names(get_name(r1), get_name(r2));
    }

    return rank;
}

Sp.lmanager.getTimerLabel = function(m) {
    if (!Ext.isNumber(m)) {
        return '';
    }
    if (m < 60) {
        return m + ' min';
    } else {
        var hours = parseInt(m / 60);
        var minutes = m - (hours * 60);
        var label = hours + ' h &nbsp;';
        if (minutes > 0) {
            label += minutes + ' min';
        }
        return label;
    }
}

Sp.lmanager.checkAccount = function(slotRec, locationRec) {
    var membershipRec = slotRec.membershipRec;
    // membershipRec is present only after adding a person
    // if the ui reloads, the slot record is reloaded.
    // in this case we left the slot as is, the verification has
    // been done prior. The problem is if meanwhile something change
    // in the account or profile... also the slot can easily be removed
    // and readded to redo verifications...
    //
    // when billing mode is set to 'other', no verification on limits is
    // done on the payer. because checking payer would require a trip
    // to the server and will slow the ui, we asume that when setting
    // payement to another person, that person has an infinite credit
    // line

    // try: using the previous problematic and probem fields
    // is case of no membershipRec, since the check account is
    // the last check, if problematic flag is set, it must be
    // related to accounts

    if (!membershipRec) {
        Log('NO membershipRec')
        if (slotRec.data.problematic) {
            Log('USING LAST PB')
            return slotRec.data.problem;
        }
        return;
    }
    var pp = Sp.ui.data.getPersonProfile(membershipRec, locationRec);

    // do no check
    if (pp.billing_mode == 'none') {
        Log('billing mode is none')
        return;
    }

    // check catalog infos
    if (!slotRec.data.item) {
        return TR("No catalog item is set for this slot");
    }
    if (!slotRec.data.element) {
        return TR("No catalog element is set for this slot");
    }
    if (!slotRec.data.price) {
        return TR("No catalog price is set for this slot");
    }

    // get catalog infos
    var item = locationRec.LocationCatalogItems().getById(slotRec.data.item);
    if (!item) {
        return TR("Error: catalog item has been removed !");
    }

    // FIXME: what to do with catalog element ?

    var price = item.LocationCatalogPrices().getById(slotRec.data.price);
    if (!price) {
        return TR("Error: catalog price has been removed !");
    }

    // no more check is done for billing other
    if (pp.billing_mode == 'other') {
        Log('billing mode is other')
        return;
    }

    // person has buyed this item
    if (Sp.ui.data.personHasBuyedItem(membershipRec, slotRec.data.item, slotRec.data.price)) {
        Log('has buyed item')
        return;
    }

    // check balance
    if (Ext.isObject(price.data.currency)) {
        var currency_uuid = price.data.currency.uuid;
    } else {
        var currency_uuid = price.data.currency;
    }
    var balance = Sp.ui.data.getPersonBalance(membershipRec, currency_uuid);
    if (!Ext.isNumber(balance)) {
        balance = 0;
    }
    if (balance >= price.data.price) {
        Log('has > balance')
        return;
    }

    // prepaid check stops here
    if (pp.billing_mode == 'pre') {
        return TR("Insufficient funds");
    }

    // just post billing_mode gets here
    var credit_line = Ext.isNumber(pp.credit_line) ? pp.credit_line : 0;
    if (balance + credit_line < price.data.price) {
        return TR("Credit limit has been exceeded");
    }

    Log('has > credit')
    Log('balance: ' + balance)
    Log('credit: ' + credit_line)
    Log('price: ' + price.data.price)

}

Sp.lmanager.getSlotJumperName = function(slotRec, locationRec) {
    var name = '';
    if (slotRec.data.person){
        name = Sp.ui.misc.formatFullname({data:slotRec.data.person}, Data.me.data.name_order, true);
    } else if (slotRec.data.phantom){
        name = slotRec.data.phantom.name;
    } else if (slotRec.data.worker){
        var worker = locationRec.Workers().getById(slotRec.data.worker);
        if (worker){
            name = worker.data.name;
        }
    }
    return name;
}

Sp.lmanager.getSlotJumpProgram = function(loadRec, slotRec, locationRec) {
    var program = [];
    if (slotRec.data.item && slotRec.data.element){
        var item = locationRec.LocationCatalogItems().getById(slotRec.data.item);
        var element = item.LocationCatalogElements().getById(slotRec.data.element);
        program.push(Ext.String.format('{0}{1}', element.data.altitude, element.data.altitude_unit));
    }
    if (slotRec.data.worker_type){
        program.push(Data.workerTypes.getById(slotRec.data.worker_type).data.label);
    } else if (slotRec.data.jump_type){
        program.push(Data.jumpTypes.getById(slotRec.data.jump_type).data.label);
    }
    if (loadRec && loadRec.data.jumpmaster_slot == slotRec.data.uuid){
        program.push(TR("Jumpmaster"));
    }
    return program.join('&nbsp;-&nbsp;');
}

Sp.lmanager.getSlotsInfos = function(loadRec, locationRec){
    var slots_infos = {};
    var aircraft = locationRec.Aircrafts().getById(loadRec.data.aircraft);
    var slots_store = loadRec.Slots();
    slots_infos.total = aircraft.data.max_slots;
    slots_infos.min = aircraft.data.min_slots;
    slots_infos.created = slots_store.getCount();
    slots_infos.used = 0;
    slots_store.each(function(s){
        if (s.data.related_slot || s.data.person || s.data.phantom || s.data.worker){
            slots_infos.used += 1;
        } else if (slots_store.find('related_slot', s.data.uuid) != -1){
            slots_infos.used += 1;
        }
    });
    slots_infos.free = slots_infos.total-slots_infos.used;
    return slots_infos;
}

Sp.lmanager.getLoadHeader = function(loadRec, locationRec, infos){
    infos = infos || Sp.lmanager.getSlotsInfos(loadRec, locationRec);
    if (infos.used > 0){
        var header_text = infos.used + ' ' + (infos.used > 1 ? TR("Jumpers") : TR("Jumper"));
        var groups_count = 0;
        var solo_count = 0;
        var slots_counts = {};
        var slots_store = loadRec.Slots();
        slots_store.each(function(s){
            if (s.data.related_slot){
                if (!Ext.isDefined(slots_counts[s.data.related_slot])){
                    slots_counts[s.data.related_slot] = 0;
                }
                slots_counts[s.data.related_slot] += 1;
            } else {
                if (!s.data.person && !s.data.phantom && !s.data.worker && slots_store.find('related_slot', s.data.uuid) == -1){
                    return
                }
                if (!Ext.isDefined(slots_counts[s.data.uuid])){
                    slots_counts[s.data.uuid] = 0;
                }
                slots_counts[s.data.uuid] += 1;
            }
        });
        Ext.Object.each(slots_counts, function(k,v){
            if (v > 1){
                groups_count += 1;
            } else {
                solo_count += 1;
            }
        });
        if (groups_count == 0){
            header_text += Ext.String.format("&nbsp;&nbsp;({0})", TR("no Groups"));
        } else {
            if (solo_count > 0){
                header_text += Ext.String.format("&nbsp;&nbsp;({0} {1} - {2} {3})", solo_count,
                                (solo_count > 1 ? TR("Solos"): TR("Solo")), groups_count,
                                (groups_count > 1 ? TR("Groups") : TR("Group")));
            } else {
                header_text += Ext.String.format("&nbsp;&nbsp;({0} {1})", groups_count,
                                (groups_count > 1 ? TR("Groups") : TR("Group")));
            }
        }
    } else {
        var header_text = TR("No Jumpers");
    }
    return header_text;
}

Sp.lmanager.hasClearance = function(locationRec, person_uuid) {
    var clr;
    locationRec.Clearances().each(function(c){
        if (c.data.person && c.data.person.uuid == person_uuid){
            clr = c;
            return false;
        }
    });
    if (clr && clr.data.approved){
        var p = Sp.ui.misc.getClearancePeriod(clr);
        if (p.end_date){
            return Ext.Date.between(new Date(), p.start_date, p.end_date);
        } else {
            return Ext.Date.isEqual(Ext.Date.clearTime(new Date()), Ext.Date.clearTime(p.start_date, true));
        }
        
    }
}

Sp.lmanager.getLoadWeight = function(locationRec, loadRec) {
    return '0 Kg';
}

File: data_server/utils/misc.py
# Copyright 2012, Nabil SEFRIOUI
#
# This file is part of Skyproc.
#
# Skyproc is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as 
# published by the Free Software Foundation, either version 3 of 
# the License, or any later version.
#
# Skyproc is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public 
# License along with Skyproc. If not, see <http://www.gnu.org/licenses/>.

import uuid
import logging
import ujson
import random
import time
import datetime
from django.db import models
from django.conf import settings

def validate_uuid(uuid_str):
    if isinstance(uuid_str, dict) and uuid_str.has_key('uuid'):
        uuid_str = uuid_str['uuid']
    try: return uuid.UUID(uuid_str).hex
    except Exception, e:
        msg = 'Invalid UUID: %s' % e
        if settings.DEBUG:
            msg += ' (%s)' % uuid_str
            logging.debug(msg)
        else:
            logging.error(msg)
        raise Exception, msg

def formatFullname(person, name_order=None, capitalize=False):
    if name_order is None:
        name_order = person.name_order
    if capitalize:
        try: first_name = person.first_name[0].upper() + person.first_name[1:].lower()
        except: first_name = person.first_name
        last_name = person.last_name.upper()
    else:
        first_name = person.first_name
        last_name = person.last_name
    if name_order == 'FL':
        return first_name + ' ' + last_name
    else:
        return last_name + ' ' + first_name

def get_default_price_for_currency(item, currency):
    try: return item.locationcatalogprice_set.filter(deleted=False, default=True, currency=currency)[0]
    except: pass

def get_default_price(membership, item):
    price = None
    if price is None and membership.override_profile and membership.currency:
        price = get_default_price_for_currency(item, membership.currency)
    if price is None and membership.profile and membership.profile.currency:
        price = get_default_price_for_currency(item, membership.profile.currency)
    if price is None and membership.person.default_currency:
        price = get_default_price_for_currency(item, membership.person.default_currency)
    if price is None and membership.location.default_currency:
        price = get_default_price_for_currency(item, membership.location.default_currency)
    return price

def get_person_profile(membership):
    ret = {}
    ret['billing_mode'] = 'pre'
    ret['currency'] = None
    ret['credit_line'] = 0
    ret['default_item'] = None
    ret['default_price'] = None
    ret['default_price'] = None
    ret['catalog_access'] = False
    ret['available_items'] = {}
    
    # billing mode
    if membership.override_profile and membership.billing_mode:
        ret['billing_mode'] = membership.billing_mode
    elif membership.profile:
        ret['billing_mode'] = membership.profile.billing_mode
    
    # currency
    if membership.override_profile and membership.currency:
        ret['currency'] = membership.currency
    elif membership.profile and membership.profile.currency:
        ret['currency'] = membership.profile.currency
        
    # credit line
    if membership.override_profile and membership.credit_line:
        ret['credit_line'] = membership.credit_line
    elif membership.profile and membership.profile.credit_line:
        ret['credit_line'] = membership.profile.credit_line
    
    # payer
    if ret['billing_mode'] == 'other':
        if membership.override_profile and membership.bill_person:
            ret['bill_person'] = membership.bill_person
        elif membership.profile and membership.profile.bill_person:
            ret['bill_person'] = membership.profile.bill_person
        else:
            ret['billing_mode'] = 'post'
    
    # catalog access
    ret['catalog_access'] = (membership.override_profile and membership.catalog_access) \
                            or (membership.profile and membership.profile.catalog_access)
    
    # default item
    if membership.override_profile and membership.default_catalog_item:
        ret['default_item'] = membership.default_catalog_item
        ret['default_price'] = membership.default_catalog_price
    elif membership.profile and membership.profile.default_catalog_item:
        ret['default_item'] = membership.profile.default_catalog_item
        ret['default_price'] = membership.profile.default_catalog_price

    # available items
    items = {}
    if membership.profile:
        for i in membership.profile.profilecatalog_set.filter(deleted=False):
            if i.price: items[i.item.uuid] = i.price
            else:
                def_price = get_default_price(membership, i.item)
                if def_price: items[i.item.uuid] = def_price
    for i in membership.membershipcatalog_set.filter(deleted=False):
        if i.price: items[i.item.uuid] = i.price
        else:
            def_price = get_default_price(membership, i.item)
            if def_price: items[i.item.uuid] = def_price
    ret['available_items'] = items
    
    # extra items
    extra_items = {}
    if membership.profile:
        for i in membership.profile.profileextracatalog_set.filter(deleted=False):
            if not extra_items.has_key(i.item.uuid):
                extra_items[i.item.uuid] = {}
                extra_items[i.item.uuid]['item'] = i.item
            if i.price: extra_items[i.item.uuid]['price'] = i.price
            else: extra_items[i.item.uuid]['price'] = get_default_price(membership, i.item)
    for i in membership.membershipextracatalog_set.filter(deleted=False):
        if not extra_items.has_key(i.item.uuid):
            extra_items[i.item.uuid] = {}
            extra_items[i.item.uuid]['item'] = i.item
        if i.price: extra_items[i.item.uuid]['price'] = i.price
        else: extra_items[i.item.uuid]['price'] = get_default_price(membership, i.item)
    ret['extra_items'] = extra_items.values()
    
    return ret

def update_prices(price, prices):
    if not price or price == 'N/A':
        return
    price = ujson.decode(price)
    for k,v in price.items():
        if not prices.has_key(k):
            prices[k] = 0
        prices[k] += v

def get_tmp_link():
    link = ''
    for i in range(2):
        link += str(uuid.uuid4()).replace('-','')
    return link[:48]

def fake_processing(a, b):
    for i in range(random.randint(a, b)):
        time.sleep(random.random())
    
def is_clear_member(membership):
    if not membership.approved:
        return
    if not membership.location.use_clearances:
        return True
    try: clr = models.get_model(settings.DATA_APP, 'Clearance').objects.get(location=membership.location, person=membership.person)
    except: return
    if not clr.approved:
        return
    if clr.end_date:
        end_date = clr.end_date
    elif clr.duration:
        if clr.unit == 'd': period = datetime.timedelta(days=1)
        elif clr.unit == 'w': period = datetime.timedelta(weeks=1)
        elif clr.unit == 'm': period = datetime.timedelta(days=30)
        elif clr.unit == 'y': period = datetime.timedelta(days=365)
        else: period = datetime.timedelta()
        end_date = clr.start_date + period
    else:
        end_date = clr.start_date
    return end_date >= datetime.date.today()
    
def check_member_account(membership, p, item, price, recurse=True):    
    if p['billing_mode'] == 'none': return True
    if p['billing_mode'] == 'other' and recurse:
        try:
            payer_membership = models.get_model(settings.DATA_APP, 'LocationMembership').objects.get( \
                                                location=membership.location, person=p['bill_person'], deleted=False)
            if not is_clear_member(payer_membership): raise Exception 
            return check_member_account(payer_membership, get_person_profile(payer_membership), item, price, recurse=False)
        except: pass
    # check if member has enough balance
    try:
        account = models.get_model(settings.DATA_APP, 'Account').objects.get( \
                    membership=membership, currency=price.currency, deleted=False, balance__gte=price.price)
        return True
    except: pass
    return p['billing_mode'] == 'post' and p['credit_line'] >= price.price and p['currency'] == price.currency
